### PR TITLE
Frontend perf improvements

### PR DESCRIPTION
**Pull request by brianlovin**
**Status**
- [ ] WIP
- [x] Ready for review
- [ ] Needs testing

**Deploy after merge (delete what needn't be deployed)**
- api
- hyperion (frontend)

Adds a secondary index to make certain queries faster, fixes a couple queries to use indexes instead of large filters, and adds pagination to `usersCommunityConnection` (to speed up loading times on power user profiles)

One thing I tried exploring in this branch @mxstbr was making it so that thread feeds load faster by not fetching as much data (e.g. is it possible to *only* fetch the data needed for the thread items in the feed). Turns out it makes the rest of the experience janky when using the inbox view, because we rely so heavily on the Apollo cache to make the inbox show threads instantly in the right column...

Anyways - please review this PR thoroughly whenever you have time :)

### CODE REVIEW

**Comment by brianlovin** [comment]
So regarding "and adds pagination to usersCommunityConnection" - turns out this is pretty damn hard to do when it's a logged in user viewing another user's profile.

The logic is working right now for: 
- signed out user viewing a user profile
- current user viewing their own profile

I'll need to think a bit harder how to do the case of a logged in user viewing another user. I can get it to paginate, just not sorted by reputation :P

File: api/models/channel.js [hunk 1]
```diff
@@ -80,23 +80,14 @@ const getChannelsByUserAndCommunity = async (communityId: string, userId: string
...
-      .without({ left: ['id', 'channelId', 'userId', 'createdAt'] })
-      // zip the tables
-      .zip()
-      // ensure we don't return any deleted channels
-      .filter(channel => db.not(channel.hasFields('deletedAt')))
-      .run()
-  );
+  return db
+    .table('usersChannels')
+    .getAll([userId, 'member'], { index: 'userIdAndRole' })
```
Comment (mxstbr) [inline api/models/channel.js:85]: We need to fetch channels a user is a moderator and an owner in here too, right?

```suggestion
    .getAll([userId, 'member'], [userId, 'moderator'], [userId, 'owner'], { index: 'userIdAndRole' })
```

File: api/models/channel.js [hunk 2]
```diff
@@ -432,22 +423,12 @@ const setMemberCount = (
...
-    .count()
-    .run();
-};
-
 const getChannelsOnlineMemberCounts = (channelIds: Array<string>) => {
   return db
     .table('usersChannels')
-    .getAll(...channelIds, {
-      index: 'channelId',
+    .getAll(...channelIds.map(id => [id, 'member']), {
```
Comment (mxstbr) [inline api/models/channel.js:429]: We should also fetch mods and owners here, not sure how to change that `.map` though to make that work… :sweat_smile: Maybe not the biggest deal?

File: api/models/thread.js [hunk 3]
```diff
@@ -94,9 +94,9 @@ export const getThreadsByChannels = (
     .table('threads')
     .getAll(...channelIds, { index: 'channelId' })
     .filter(thread => db.not(thread.hasFields('deletedAt')))
-    .orderBy(...order)
     .skip(after || 0)
     .limit(first || 999999)
+    .orderBy(...order)
```
Comment (mxstbr) [inline api/models/thread.js:99]: I'm 99% sure this needs to stay before the `.skip` and `.limit`. Otherwise we're fetching the first n records, but they might not actually be the first n records of the order we want.

For example, imagine the records returned from the database are:

```JS
[{
  id: 'asdf-1',
  createdAt: '2018-11-11'
}, {
  id: 'asdf-2',
  createdAt: '2018-12-01'
}]
```

Note how `asdf-2` was posted after `asdf-1`. If we do a `.first(1).orderBy(db.desc('createdAt'))`, we will get the thread `asdf-1` even though we should've gotten the the thread `asdf-2` because we truncated the list before ordering it. 

If we do the `orderBy` before the truncation we will correctly get the `asdf-2` record. Does that make sense?

I think we could add an index to threads called `channelIdAndTrending`, and then use the `channelIdAndLastActive` or `channelIdAndTrending` indexes here to efficiently order after a `.between`, right? I can't off the top of my head figure out how we'd do a `.between` on multiple `channelIds`, but I think it should be possible?

File: api/queries/channel/memberConnection.js [hunk 4]
```diff
@@ -29,7 +29,7 @@ export default async (
     .then(users => loaders.user.loadMany(users))
     .then(result => ({
       pageInfo: {
-        hasNextPage: result && result.length >= first,
+        hasNextPage: result && result.length > first,
```
Comment (mxstbr) [inline api/queries/channel/memberConnection.js:32]: I don't think this works here. This would only work if we fetch `after + 1` records to see if there is any records left after the one we load and then not send that `+ 1` record down. As-is, I'm pretty certain this doesn't work correctly.

The original logic only has a single edge case where it breaks: if the last page has the exact number of records that were requested, this query will think there's a next page and we'll show the loading indicator on the frontend while the non-existant next page is fetched. Then it'll return `false` (because there is no next page, and `result.length === 0`) and everything will go back to normal.

In my head that's a perfectly valid tradeoff, showing a loading indicator in that specific case a couple hundred milliseconds for too long instead of complicating this resolver and query, and it's how we've handled this in almost all connection resolvers (with the notable exception of thread connections where we do fetch `after + 1` records).

I say we go back to the original logic? Otherwise we'd have to rework this to fetch `after + 1` records and then discard the last record, which is too much work imo.

```suggestion
        hasNextPage: result && result.length >= first,
```

File: api/queries/community/memberConnection.js [hunk 5]
```diff
@@ -50,7 +50,7 @@ export default async (root: DBCommunity, args: Args, ctx: GraphQLContext) => {
     .then(users => loaders.user.loadMany(users))
     .then(result => ({
       pageInfo: {
-        hasNextPage: result && result.length >= first,
+        hasNextPage: result && result.length > first,
```
Comment (mxstbr) [inline api/queries/community/memberConnection.js:53]: Same as above


```suggestion
        hasNextPage: result && result.length >= first,
```

File: api/queries/user/communityConnection.js [hunk 6]
```diff
@@ -1,39 +1,66 @@
...
 
-  let communities;
+  const cursor = decode(after);
+  // Get the index from the encoded cursor, asdf234gsdf-2 => ["-2", "2"]
+  const lastDigits = cursor.match(/-(\d+)$/);
+  const lastCommunityIndex =
+    (lastDigits && lastDigits.length > 0 && parseInt(lastDigits[1], 10)) || 0;
+
+  let communities,
+    forceNoPagination = false;
```
Comment (mxstbr) [inline api/queries/user/communityConnection.js:28]: Why is this here and why is it hard-coded? Do we need this?

File: api/types/User.js [hunk 7]
```diff
@@ -106,19 +107,22 @@ const User = /* GraphQL */ `
...
+    ): UserCommunitiesConnection!
     channelConnection: UserChannelsConnection!
     directMessageThreadsConnection(
       first: Int = 15
       after: String
     ): UserDirectMessageThreadsConnection!
       @cost(complexity: 1, multipliers: ["first"])
     threadConnection(
-      first: Int = 20
+      first: Int = 10
```
Comment (mxstbr) [inline api/types/User.js:121]: Very good idea. 👍

File: api/types/User.js [hunk 8]
```diff
@@ -106,19 +107,22 @@ const User = /* GraphQL */ `
...
       after: String
       kind: ThreadConnectionType
     ): UserThreadsConnection! @cost(complexity: 1, multipliers: ["first"])
-    everything(first: Int = 20, after: String): EverythingThreadsConnection!
+    everything(first: Int = 10, after: String): EverythingThreadsConnection!
```
Comment (mxstbr) [inline api/types/User.js:125]: 👍

**Review by mxstbr — CHANGES_REQUESTED** [review]
Overall love the `userIdAndIsParticipant` index and the reduction of the default fetching amount 👍 

Have some comments inline about some backend query stuff, overall it's great though, thank you so much for digging into this!

File: api/models/channel.js (hunk 1 above)
Comment (brianlovin) [reply api/models/channel.js:85]: I guess again, I'm confused by this index because in our database if you are a moderator or an owner you are *also* a member. So shouldn't fetching `[userId, 'member']` always return members including the team?

File: api/models/thread.js (hunk 3 above)
Comment (brianlovin) [reply api/models/thread.js:99]: Good call - I'm dumb. Let's tackle this one again later then.

**Comment by brianlovin** [comment]
Your feedback here is *super* useful - I made a lot of mistakes! I'm going to scrap this branch since it's harder to tweak apart the good bits and bad bits; I have a followup PR with the good bits I'll submit today.

File: api/models/channel.js (hunk 1 above)
Comment (mxstbr) [reply api/models/channel.js:85]: No, it can only have a single value, you're either-or

Comment (brianlovin) [reply api/models/channel.js:85]: I guess I'm saying won't the db lookup return the proper results regardless?

Comment (mxstbr) [reply api/models/channel.js:85]: No. The index only has a single value, so you're **either** a "member" **or** a "moderator" **or** an "owner". Members are exclusive of moderators are exclusive of owners, these groups have 0 overlap in the index.
